### DISCUSSION TITLE
How should I instantiate a Go struct to return it in a method and then use it on GDScript?

### DISCUSSION THREAD

**Discussion by pafuent** [quaadgras/graphics.gd#35] [Q&A] [answered]
I have this Go code...
```go
type Bar struct {
	classdb.Extension[Bar, Object.Instance]

	Message string
}

type Foo struct {
	classdb.Extension[Foo, Object.Instance]
}

func (k *Bar) GetBar(message string) Object.Instance {
        bar := &Bar{Message: message}
	return Object.Instance(bar.AsObject())
}
```
...then on a GDScript I create a Foo instance and call GetBar...
```
	var foo: Foo = Foo.new()
	var bar = foo.GetBar("Hello!!!")
	print(bar.Message)
```
...but when I run it a got this error: `SCRIPT ERROR: Invalid get index 'Message' (on base: 'null instance').`

Reading the code I think the problem is [here](https://github.com/grow-graphics/gd/blob/master/internal/ctx.go#L55). This line always return true due to the previous line, so I always get `native.(T)` instead of `ptr`

If I change `GetBar` to be like this...
```go
func (k *Bar) GetBar(ctx gd.Context, message gd.String) gd.Object { // outdated
	ptr := new(Bar)
	object := ctx.API.ClassDB.ConstructObject(ctx, ctx.StringName(strings.TrimPrefix(reflect.TypeOf(ptr).Elem().Name(), "class")))
	object.Set(ctx.StringName("Message"), ctx.Variant(message))
	ptr.SetPointer(object.AsPointer())
	return ptr.AsObject()
}
```
...it's works as I expect.

Also I'm wondering how I should write `GetBar` to not call `object.Set(ctx.StringName("Message"), ctx.Variant(message))` to set the value of `Message` on `bar`. In other words: Is there a way to set `Message` on Go `Bar` instance instead of using `Object.Set`?

**Comment by Splizard** [comment]
I would expect this to work:
```go
func (k *Bar) GetBar(message string) Object.Instance {
	var result = new(Bar)
        result.Message = message
        return Object.Instance(result.AsObject())
}
```

**Reply by pafuent** [reply]
I tried that, but returns the error that I mentioned before: `SCRIPT ERROR: Invalid get index 'Message' (on base: 'null instance').`

**Reply by pafuent** [reply]
I got it working using `Pin()`
```go
func (f *Foo) GetBar(ctx gd.Context, message gd.String) gd.Object {
	if f.ctx.API == nil {
		f.ctx = f.Pin()
	}
	var result = gd.Create(f.ctx, new(Bar))
	result.Message = message.Copy(f.ctx)
	return result.AsObject()
}
```
If I got it correctly, what I'm doing here is to tie `Bar` instance to a context that will be hold and handled by `Foo`.

**Reply by Splizard** [reply, marked as answer]
@pafuent I have made some changes around `gd.Context` to hopefully improve the experience here, you can do the following now and I can confirm this works with the latest master:
```go
func (k *Bar) GetBar(message string) Object.Instance {
	var result = new(Bar)
        result.Message = message
        return Object.Instance(result.AsObject())
}
```

**Reply by TeddyDD** [reply]
Does it needs to be returned as gd.Object or can you return the struct directly?

**Reply by Splizard** [reply]
At the moment, a `gd.Object` but in future, other values will be supported.

**Comment by jugglingjsons** [comment]
Hey! 
Not sure if I should post here but I've just tried running the `eg/1d/signals` and it doesn't seem to work.

I am getting `SCRIPT ERROR: Invalid call. Nonexistent function 'DoSomething' in base 'Signals'.`

I believe it's using similar approach to the one described here. Trying out the mvp from this thread now.

Same results, can't seem to find it. I have a feeling it's connected to what I've done with the extension for macos. Not sure how to verify that though. Could someone please check on a different system ? 

@Splizard could you try to verify the above ? I am actually not sure if it should be working currenlty

**Reply by Splizard** [reply]
Hey @jugglingjsons, I had an issue opened recently #92 about exports and the GDScript naming convention being snake_case, so they should now be getting exported to follow the convention by default.

I think this affects signals too, so if there is no `gd` tag, then `DoSomething` will become `do_something` using `String.ToSnake`, I'll fix the signals example but have you run into this as an issue?

**Reply by jugglingjsons** [reply]
I was specifically trying this part:

```
func (k *Bar) GetBar(message string) Object.Instance {
	var result = new(Bar)
        result.Message = message
        return Object.Instance(result.AsObject())
}
```

After correcting the case references it seems to work as it gets into the code ,  however now it throws with:

`panic: gdextension: unsupported Go -> Godot type gd.Object`

I am trying to offload expensive logic to go, and pass results as data and am wondering how to do it efficiently for big objects.

**Reply by Splizard** [reply]
@jugglingjsons there does seem to be an issue here, I've pushed a fix and can confirm that GetBar is working in the example project. You should also be able to return the Go pointer type ie. `*Bar` instead of `Object.Instance`.

https://github.com/grow-graphics/eg/blob/master/1d/original/main.go#L38

**Reply by jugglingjsons** [reply]
Beautiful, I admire your responsiveness a lot!
The only issue I've found is when I tried to return an array / non primitive it would in theory make the data available, but the program would exit with malalloc. 
I've done this roughlty:

```
type Vector2D struct {
	X float64
	Y float64
}
type Bar struct {
	classdb.Extension[Bar, Object.Instance]

	Message string
	Vecs    []Vector2D
}

```

Truth be told I don't think I need to pass it this way anyways :P 
I assume if I set custom_property from go on some node, I will be able to access it form gdscript anyways - as this is the end goal I am aiming for :)

**Reply by Splizard** [reply]
I mean, this should work... I think the goal now  is to have almost any Go value you want in your class extension and `gd` will try and export it in the way that makes the most sense with gdscript. Are you expecting this to be copied across as a `PackedVector2Array` ? It's probably trying to transform this into an `Array[Dictionary]` because the godot vector types have float32 elements by default and gd isn't currently looking at this type and thinking "hey this is actually a Vector2".

**Reply by jugglingjsons** [reply]
It does work to an extent. 
It usually runs fine and the values are available in both GDScript and Go.

However sometimes upon exit through SIGKILL it throws with:

> Hello Property from Godot -> Go -> Godot
objc[12493]: Hash table corrupted. This is probably a memory error somewhere. (table at 0x1f9ce75c8, buckets at 0x13f13ec40 (0 bytes), 4 buckets, 1 entries, 0 tombstones, data 0x0 0x0 0x0 0x0)
Error: signal: abort trap

It also sometimes exits on it's own with: 

```
godot(14595,0x1f9ce7ac0) malloc: Corruption of free object 0x123eaed80: msizes 9/10 disagree
godot(14595,0x1f9ce7ac0) malloc: *** set a breakpoint in malloc_error_break to debug
Error: signal: abort trap

main.wrap()
        /Users/danieljadczak/go/pkg/mod/graphics.gd@v0.0.0-20250206013510-029f470fbe0a/cmd/gd/main.go:412
```

With not changes to files.

Both of these errors do not happen always - as in I can reproduce them, but not every time even when following the exact same actions. The whole code I am using is here:

```
	var bar = hello.get_bar("Hello Bar")

	print(bar.vecs.size())
	print(bar.vecs)
	for v in bar.vecs:
		print(v)
		print(v.X)
```

and 

```
type Vector2D struct {
	X float64
	Y float64
}
type Bar struct {
	classdb.Extension[Bar, Object.Instance]

	Message string
	Vecs    []Vector2D
}

func (h *HelloWorld) Bar(message string) *Bar {
	return &Bar{
		Message: message,
		Vecs:    []Vector2D{{1, 2}, {3, 4}},
	}
}
```

I know it's a little artificial, and perhaps Mac specific, or it's some compiler cache perhaps ? Not sure I know way too little and there is a lot of indirection in terms of pointer ownership here :) 

Anyways, as mentioned it's not that I need this to work 100%.
There's no point in creating artificial data transport layer when one can already access the data exchange through Go -> GDscript node patterns which should work just fin.

I will try to give it a go over the weekend with a proper example of what I want to do which is passing data from Go to  MultimeshInstance.custom_data and see if it works. If it does, it's all gucci ;)

**Reply by Splizard** [reply]
@jugglingjsons thanks for the report, I'm not sure that I can reproduce this, if you run into this again, can you provide some more details about your `go version` and `uname -a` ?

**Reply by jugglingjsons** [reply]
Hey , sure thing:
> go version go1.23.5 darwin/arm64

and 

> Darwin Daniels-MacBook-Pro.local 23.4.0 Darwin Kernel Version 23.4.0: Fri Mar 15 00:10:42 PDT 2024; root:xnu-10063.101.17~1/RELEASE_ARM64_T6000 arm64

**Reply by Splizard** [reply]
@jugglingjsons thanks, have been able to reproduce and have pushed some fixes for this issue.
